subsys/bluetooth/host/hci_core.c

```diff
@@ -364,31 +365,41 @@ int bt_hci_cmd_send_sync(u16_t opcode, struct net_buf *buf,
 	net_buf_put(&bt_dev.cmd_tx_queue, buf);
 
 	err = k_sem_take(&sync_sem, HCI_CMD_TIMEOUT);
-	__ASSERT(err == 0, "k_sem_take failed with err %d", err);
+	if (err) {
+		BT_WARN("Failed to send command to driver (%d)", err);
```

Comment:
This should be `BT_ERR()` (sorry if I missed it previously). Basically, `BT_ERR()` should be used when there's a definitive error condition coming from outside of our subsystem or component (the HCI driver or controller in this case). `BT_WARN()` OTOH is for cases where there's not necessarily an error, but might be, and it's important to make the developer/user aware of this. `BT_INFO()` should be reserved for rare circumstances where we have some generally informational content to show the developer/user - e.g. that's why it's used to log the basic controller information upon HCI init.

Reply:
Actually, putting the logging details aside: why did you remove the assert and try to handle this at runtime? Isn't the HCI command flow control broken if we fail to wait for the command complete/status? IIRC that's why this was an assert.

Reply:
IIRC originally this was with K_FOREVER, so neither assert nor BT_ERR() was useful. We later introduced the HCI_CMD_TIMEOUT so we could log this condition, but it doesn't change the fact that something is critically broken if this happens and we probably shouldn't attempt to continue normal operation of the stack.

Reply:
Also, the log message is incorrect: we *did* succeed sending the command to the driver, but then subsequently timed out waiting for a command complete/status.

Reply:
At first I wanted to add an BT_ASSERT instead of __ASSERT, since as you say something is broken, and the __ASSERT can be disabled. However I thought it was too harsh to crash the host in a split build where you are communicating with an HCI controller that might have crashed.

But you are right, something is critically broken. Maybe we could unset the BT_DEV_READY flag, send an HCI reset command, but unsure if that handles everything.

Maybe a BT_ASSERT is the best option after all.

Reply:
yeah, I'd go with that as well

Reply:
Added BT_ASSERT_MSG to keep the message.